zookeeper-server/src/test/java/org/apache/zookeeper/server/quorum/LearnerTest.java

```diff
@@ -300,4 +307,52 @@ public void syncTest() throws Exception {
         }
     }
 
+    @Test
+    public void truncFailTest() throws Exception {
+        final boolean[] exitProcCalled = {false};
+
+        ServiceUtils.setSystemExitProcedure(new Consumer<Integer>() {
+            @Override
+            public void accept(Integer exitCode) {
+                exitProcCalled[0] = true;
+                assertThat(exitCode, equalTo(ExitCode.QUORUM_PACKET_ERROR.getValue()));
```

Comment:
This assertion will not happen in the main junit thread, so if it asserts false, the behaviour will be unpredictable (the test will probably ignore it). 

Instead of this, I'd replace exitProcCalled with a ```CompletableFuture<Integer>``` and at the end of the  test method do
```
assertThat(exitCode.get(10, TimeUnit.SECONDS), equalTo(ExitCode.QUORUM_PACKET_ERROR.getValue()));
```

Reply:
I double checked it. Consumer's accept() method is called synchronously, no threading involved. In other words it will be called on junit's main thread.

Reply:
Ok. I see that now.